processor/abci.go

```diff
@@ -601,6 +615,17 @@ func (app *App) OnBeginBlock(req tmtypes.RequestBeginBlock) (ctx context.Context
 	app.log.Debug("entering begin block", logging.Time("at", time.Now()), logging.Uint64("height", uint64(req.Header.Height)))
 	defer func() { app.log.Debug("leaving begin block", logging.Time("at", time.Now())) }()
 
+	// if there is an approved protocol upgrade proposal and the current block height is later than the proposal's block height then take a snapshot and wait to be killed by the process manager
+	if app.protocolUpgradeService.TimeForUpgrade() {
+		app.snapshot.SnapshotNow(ctx)
```

Comment:
I don't think we can take the snapshot here. Assuming I follow properly we are at block `n`, the first block that will be in the new version, and effectively snapshoting block`n-1`, the last block of the old version.

When we pull down the old core process and bring up the new one with the snapshot data from block `n-1` tendermint will want the commit hash of block `n-1` during the abci `Info` call, which we won't have.  So I think you'll have to move the forcing of the snapshot to `OnCommit` of block `n-1`.